### DISCUSSION TITLE
[API Proposal]: Collection Expressions and ```String.Empty```

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#8316] [Language Ideas]
### Background and motivation

It would be really convenient if ``` [] ``` could be used as a shorthand for ```String.Empty```.

### API Proposal

I am not sure what is necessary to make this possible.

### API Usage

```csharp
public record Person {
  public string Name { get; init; } = [];
}
```

### Alternative Designs
We could leave it as-is.
Also, someone could just use ```""``` but that isn't as nice as [].

**Comment by alrz** [comment]
If anything, wouldn't it make sense to emit "" as string.empty? 

That's assuming that this has any advantage over it (on that, there's https://github.com/dotnet/csharplang/discussions/8300)

**Comment by julealgon** [comment]
> It would be really convenient if `[]` could be used as a shorthand for `String.Empty`.

How so?

> Also, someone could just use `""` but that isn't as nice as `[]`.

Why? How are you defining "niceness" here?

**Reply by HaloFour** [reply]
I agree, I think `""` is a lot easier to reason about than `[]` in the context of a `string`.  I don't think it's as common to consider a `string` to be equivalent to a `char[]` so much.

**Reply by KennethHoff** [reply]
I find `""` undesirable because I can never truly know if it's empty or if it contains zero-width characters (unless you're using a reader that makes that clear, which I don't know if GitHub does)

**Reply by HaloFour** [reply]
That seems a bit extreme, and if you're really trying to guard against that then `string.Empty` should be sufficient.

I'm not opposed to `[]` being considered an empty string, but not on it's own.  I don't think that's worth doing without allowing collection literals to be used to create strings in general, e.g. `string foo = [ 'h', 'e', 'l', 'l', 'o' ];`.

**Reply by KennethHoff** [reply]
> That seems a bit extreme, and if you're really trying to guard against that then `string.Empty` should be sufficient.

I do exclusively use `string.Empty` already :)

**Reply by jnm2** [reply]
I'd suggest an analyzer that shows warning or info squiggles for strings that could be confused with the empty string. The problem wouldn't be limited just to the empty string literal, either.

**Reply by colejohnson66** [reply]
Honestly, an analyzer that flags any non-visible or formatting Unicode codepoints would be nice. If I need a non-breaking space or zero-width joiner, I define it as a constant and use it with interpolation:

```csharp
private const char NBSP = '\xA0';
private const char ZWJ = '\x200D';
public const string SomeString = $"Start{NBSP}Middle{ZWJ}End";
```
